### ISSUE TITLE
Incorrect percent encoding under Sonoma

### ISSUE THREAD

**Issue by low-batt** [iina/iina#4861]
**System and IINA version:**

- macOS 14.4
- IINA 1.3.4

**Expected behavior:**
IINA correctly percent encodes URLs.

**Actual behavior:**
When running under macOS Sonoma IINA will perform percent encoding _twice_ resulting in malformed URLs. @xfoxfu provided two examples:

Opening this URL in Firefox:
```text
iina://weblink?url=https://example.com/%5Bfoo%20bar.mkv
```

IINA changes `%5Bfoo%20bar.mkv` to be `%5Bfoo%2520bar.mkv`:
```text
13:54:42.969 [iina][d] Parsing URL iina://weblink?url=https://example.com/%5Bfoo%20bar.mkv
13:54:42.969 [player0][d] Open URL: https://example.com/%5Bfoo%2520bar.mkv
13:54:42.969 [player0][d] Opening https://example.com/%5Bfoo%2520bar.mkv in main window
```

Opening this URL in Firefox:
```text
iina://weblink?url=https://example.com/%25foo%20bar.mkv
```

IINA changes `%25foo%20bar.mkv` to be `%25foo%2520bar.mkv`:
```text
13:51:00.896 [iina][d] Parsing URL iina://weblink?url=https://example.com/%25foo%20bar.mkv
13:51:00.897 [player0][d] Open URL: https://example.com/%25foo%2520bar.mkv
13:51:00.897 [player0][d] Opening https://example.com/%25foo%2520bar.mkv in main window
```

This is a change from the behavior under macOS Ventura.

**Steps to reproduce:**
- Enable logging in IINA
- Open one of the URLs given above in Firefox
- Look for the message shown above in IINA's log file

- [ ] MPV does not have this problem.

This is an issue with IINA code. I have not checked if `mpv` also has this problem.

**How often does this happen?**
Every time.

**Comment by low-batt** [comment]
### Analysis

@xfoxfu Traced the change in behavior to an intentional change by Apple. From the [Xcode 15 Release Notes](https://developer.apple.com/documentation/xcode-release-notes/xcode-15-release-notes#Resolved-Issues):
> - Fixed: For apps linked on or after iOS 17 and aligned OS versions, URL parsing has updated from the obsolete RFC 1738/1808 parsing to the same RFC 3986 parsing as URLComponents. This unifies the parsing behaviors of the URL and URLComponents APIs. Now, URL automatically percent- or IDNA-encodes invalid characters to help create a valid URL.
> 
> 	To check if a `urlString` is strictly valid according to the RFC, use the new `URL(string: urlString, encodingInvalidCharacters: false)` initializer. This init leaves all characters as they are and will return `nil` if `urlString` is explicitly invalid. (93368104)

Many thanks to @xfoxfu for identifying the cause of the change in behavior.

Opening this URL in Firefox under macOS Ventura:
```text
iina://weblink?url=https://example.com/%5Bfoo%20bar.mkv
```

IINA correctly forms the URL:
```text
14:26:32.836 [iina][d] Parsing URL iina://weblink?url=https://example.com/%5Bfoo%20bar.mkv
14:26:32.837 [player0][d] Open URL: https://example.com/%5Bfoo%20bar.mkv
14:26:32.837 [player0][d] Opening https://example.com/%5Bfoo%20bar.mkv in main window
```

That means this problem is a regression and is therefore a high priority to fix.

_Unfortunately_…

Opening this URL in Firefox under macOS Ventura:
```text
iina://weblink?url=https://example.com/%25foo%20bar.mkv
```

**FAILS** reporting the URL can not be percent encoded:
```text
14:25:51.188 [iina][d] Parsing URL iina://weblink?url=https://example.com/%25foo%20bar.mkv
14:25:51.188 [player0][e] Cannot add percent encoding for https://example.com/%foo bar.mkv
```

There is a second, already existing, problem. Adding additional logging:
```swift
      Logger.log("Percent encoding \(str)")
      guard let pstr = str.addingPercentEncoding(withAllowedCharacters: .urlAllowed) else {
        Logger.log("Cannot add percent encoding for \(str)", level: .error, subsystem: subsystem)
        return
      }
      Logger.log("Percent encoded \(pstr)")
      guard let url = URL(string: pstr) else {
        Logger.log("Cannot parse url for \(pstr)", level: .error, subsystem: subsystem)
        return
      }
```

Shows the `Cannot add percent encoding` error is bogus:
```text
14:29:51.383 [iina][d] Parsing URL iina://weblink?url=https://example.com/%25foo%20bar.mkv
14:30:03.043 [iina][d] Percent encoding https://example.com/%foo bar.mkv
14:30:03.048 [iina][d] Percent encoded https://example.com/%foo%20bar.mkv
14:30:03.048 [player0][e] Cannot parse url for https://example.com/%foo%20bar.mkv
```

The `addingPercentEncoding` method completed successfully. It was the URL initializer that failed.

Why did IINA not report this same error running under Sonoma?

Looking at the log message from the test under Sonoma shows `%foo%20bar.mkv` was changed to `%5Bfoo%2520bar.mkv`:
```text
13:54:42.969 [player0][d] Opening https://example.com/%5Bfoo%2520bar.mkv in main window
```
The URL initializer did not fail because it applied percent encoding to the `%` characters. Why did that not happen under Ventura? Looking at the call to `addingPercentEncoding`:

```swift
guard let pstr = str.addingPercentEncoding(withAllowedCharacters: .urlAllowed), let url = URL(string: pstr) else {
```

It uses `urlAllowed` which does not require `%` to be encoded:
```swift
extension CharacterSet {
  static let urlAllowed: CharacterSet = {
    var set = CharacterSet.urlHostAllowed
      .union(.urlUserAllowed)
      .union(.urlPasswordAllowed)
      .union(.urlPathAllowed)
      .union(.urlQueryAllowed)
      .union(.urlFragmentAllowed)
    set.insert(charactersIn: "%")
    return set
  }()
}
```

Percent was added to the set a long time ago in commit 84927d760099deaaba9eae9bd8d74f9339d57db0.

The reason why [CharacterSet](https://developer.apple.com/documentation/foundation/characterset#2902136) provides the the different sets is that each URL component has different rules regarding what is allowed. It is concerning that IINA is trying to apply percent encoding to a full string. Posts on the net recommend using [URLComponents](https://developer.apple.com/documentation/foundation/urlcomponents) which has always handled percent encoding.

### Fixing

Currently PR #4860 removes the call to `addingPercentEncoding` and directly passes the string to the URL initializer. Even when running under macOS Sonoma this might fail. This test program demonstrates the problem (`.txt` added to make GitHub happy):

[URLTest.tar.gz.txt](https://github.com/iina/iina/files/14817666/URLTest.tar.gz.txt)

When built with Xcode 15 the proposed fix works:
```text
Proposed fix:
https://example.com/%5Bfoo%20bar.mkv
```

But when built with Xcode 13 the new URL initializer behavior is not activated and the fix fails:
```text
Proposed fix:
Cannot parse url for https://example.com/[foo bar.mkv
```

Full test output:
<details>
<summary><b>Xcode 15, macOS Sonoma</b></summary>

```text
Built using Xcode 15.3 and macOS SDK 14.4
Running under macOS Version 14.4 (Build 23E214)

----------------------------------------
String to open:
https://example.com/[foo bar.mkv

Current code:
Percent encoded https://example.com/[foo%20bar.mkv
https://example.com/%5Bfoo%2520bar.mkv

Proposed fix:
https://example.com/%5Bfoo%20bar.mkv

Conditional fix:
Compiled with Xcode 15+
https://example.com/%5Bfoo%20bar.mkv

----------------------------------------
String to open:
https://example.com/%foo bar.mkv

Current code:
Percent encoded https://example.com/%foo%20bar.mkv
https://example.com/%25foo%2520bar.mkv

Proposed fix:
https://example.com/%25foo%20bar.mkv

Conditional fix:
Compiled with Xcode 15+
https://example.com/%25foo%20bar.mkv
```
</details>


<details>
<summary><b>Xcode 13, macOS Sonoma</b></summary>

```text
Built using Xcode 13.1 and macOS SDK 12.0
Running under macOS Version 14.4 (Build 23E214)

----------------------------------------
String to open:
https://example.com/[foo bar.mkv

Current code:
Percent encoded https://example.com/[foo%20bar.mkv
https://example.com/%5Bfoo%20bar.mkv

Proposed fix:
Cannot parse url for https://example.com/[foo bar.mkv

Conditional fix:
Percent encoded https://example.com/[foo%20bar.mkv
https://example.com/%5Bfoo%20bar.mkv

----------------------------------------
String to open:
https://example.com/%foo bar.mkv

Current code:
Percent encoded https://example.com/%foo%20bar.mkv
Cannot create URL from https://example.com/%foo%20bar.mkv

Proposed fix:
Cannot parse url for https://example.com/%foo bar.mkv

Conditional fix:
Percent encoded https://example.com/%foo%20bar.mkv
Cannot parse url for https://example.com/%foo%20bar.mkv
```
</details>

The fix has to take into account the version of Xcode IINA was built with. Although the IINA [README](https://github.com/iina/iina?tab=readme-ov-file#using-the-pre-compiled-libraries) says:
> Open iina.xcodeproj in the [latest public version of Xcode](https://apps.apple.com/app/xcode/id497799835). _IINA may not build if you use any other version_.

IINA has plans to _try_ and create one more release that supports macOS 10.11+. That requires building with Xcode 13. So it is currently undesirable to break compatibility with older versions of Xcode.

Searching shows there are other places in the code where `addingPercentEncoding`is called before passing a string to the URL initializer. Additional fixes may be required.

The next release of IINA is supposed to be a stabilizing release with a focus on fixing regressions. As the primary problem in this issue is a regression, that needs to be fixed. I'm thinking we should hold off fixing the other problem with failing to percent encode `%` until after that release as fixing that entails some risk.

**Comment by xfoxfu** [comment]
Thanks for your detailed test. I understand the next version is a stabilizing release, so the current pull request could be aligned with the current suggestion to avoid introducing new regressions.

As a further followup, I am wondering why the URL needs to be precent-encoded, as it is observed that _mpv_ supports URL with special characters.

```
mpv 'https://example.com/[foobar] lorem ipsum.mkv'
```

Therefore, I would like to know if it is OK to assume that precent-encoding the URL is for making `URL.init(string:)` happy as it would reject URL with preserved characters before `URL.init(string:encodingInvalidCharacters:)` was introduced.

If so, I think it could be better to make precent-encoding as a fallback for `URL.init(string:)`, which looks like

```swift
func parseUrl(urlStr: String) -> URL? {
    var urlComponents = URLComponents(string: urlStr)
    if urlComponents == nil {
        guard let pUrlStr = urlStr.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) else { return nil }
        urlComponents = URLComponents(string: pUrlStr)
    }
    return urlComponents?.url
}
```

and this logic could be shared between opening URL (1) from URL scheme and (2) from UI, which also addresses inconsistent behavior between these now on macOS Ventura. I have setup a GitHub repo with Actions to check the behavior between different macOS and XCode versions: https://github.com/xfoxfu/test-iina-swift-url.

And the following test cases should be covered:

|URL|original filename|URL Scheme|
|-|-|-|
|<https://example.com/%5Bfoo%20bar.mkv>|`[foo bar.mkv`|`iina://weblink?url=https://example.com/%5Bfoo%20bar.mkv`|
|<https://example.com/%25%20.mkv>|`% .mkv`|`iina://weblink?url=https://example.com/%2525%20.mkv`|
|<https://example.com/%2520.mkv>|`%20.mkv`|`iina://weblink?url=https://example.com/%252520.mkv`|

I am not exactly sure about the encoding-decoding part, and some more time is required reading the [RFC 3986](https://www.rfc-editor.org/rfc/rfc3986) to provide precise test cases. I think this should be mostly reflecting the URL handling logic of mpv and browsers.

**Comment by low-batt** [comment]
I've created issue  #4862 for the failure to handle `%25` correctly. We should continue the discussion in that issue.

**Comment by low-batt** [comment]
The fix for this issue has been merged into the IINA `develop` branch. GitHub [automatically closed](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) the linked issue in reaction to the merge. I am reopening this issue until the fix is available in an official release of IINA so that users intending to report this problem can easily locate this existing report. Once the fix for this issue has been included in an official release this issue will be closed.

**Comment by low-batt** [comment]
This was fixed in IINA 1.3.5.
